source/extensions/filters/http/cache/cache_headers_utils.cc

```diff
@@ -168,6 +168,100 @@ absl::optional<uint64_t> CacheHeadersUtils::readAndRemoveLeadingDigits(absl::str
   return absl::nullopt;
 }
 
+absl::flat_hash_set<std::string> VaryHeader::parseAllowlist() {
+  // TODO(cbdm): Populate the hash_set from
+  // envoy::extensions::filters::http::cache::v3alpha::CacheConfig::allowed_vary_headers.
+  // Need to make sure that the headers we add here are valid values (i.e., not malformed). That
+  // way, we won't have to check this again in isAllowed.
+  return {"accept-encoding"};
+}
+
+bool VaryHeader::isAllowed(const absl::flat_hash_set<std::string>& allowed_headers,
+                           const Http::ResponseHeaderMap& headers) {
+  if (!hasVary(headers)) {
+    return true;
+  }
+
+  std::vector<std::string> varied_headers =
+      parseHeaderValue(headers.get(Http::Headers::get().Vary));
+
+  // If the vary value was malformed, it will not be contained in allowed_headers.
+  for (const std::string& header : varied_headers) {
+    if (!allowed_headers.contains(header)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool VaryHeader::hasVary(const Http::ResponseHeaderMap& headers) {
+  const Http::HeaderEntry* vary_header = headers.get(Http::Headers::get().Vary);
+  return ((vary_header) && (!vary_header->value().empty()));
```

Comment:
`return vary_header != nullptr && !vary_header->value().empty();`

that is: several sets of parens aren't necessary, but in Envoy by convention we use explicit compares to nullptr rather than using a ptr as a bool.